### PR TITLE
[2055] Add custom widget to edit EMF references

### LINKED ISSUE #2055: Add custom widget to edit EMF references
On the desktop side, Sirius (via EEF) provides rich widgets in its properties view to edit object references.
They support both mono-valued and multi-valued references, either contained or not, with all the relevant operations directly available in the widget: value selection, re-ordering, creation of new objects to add to the reference, etc.

On the web side, we only support a limited subset of this with the `Select` and `Multiselect` widgets, but by default they are only used for non-containment reference, and they do not support the richer behaviors of the EEF counterparts.

We should provide a new _custom_ widget to edit references which can be used either directly from the core APIs or via the View DSL.
They will provide the same level of configuration and features as its EEF counterpart: the specifier simply indicates which setting (owner and reference) to edit, and the widget adapts its look and features to the characteristics of the reference.

### PR DESCRIPTION
**Pull request by pcdavid**
This is a first iteration of the new reference widgets.
It will be improved (support for styling, actions on items, selction/creation modals, scalability of the UI...), but it's already functional and does not risk causing any impact on the rest, so should probably be merged soon (unless reviewers find significant issues).

### CODE REVIEW

File: package-lock.json
Comment (AxelRICHARD) [inline package-lock.json:1]: Is it normal to have package-lock.json in this PR?

File: packages/formdescriptioneditors/frontend/sirius-components-formdescriptioneditors/src/WidgetEntry.tsx [hunk 1]
```diff
@@ -503,6 +503,7 @@ export const WidgetEntry = ({
         <div
           className={classes.widgetElement}
           onClick={handleClick}
+          tabIndex={0}
```
Comment (AxelRICHARD) [inline packages/formdescriptioneditors/frontend/sirius-components-formdescriptioneditors/src/WidgetEntry.tsx:506]: What behavior this line fixes?

File: packages/forms/backend/sirius-components-widget-reference-view-edit/.project [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+			<arguments>
+			</arguments>
+		</buildCommand>
+		<buildCommand>
+			<name>org.eclipse.m2e.core.maven2Builder</name>
+			<arguments>
+			</arguments>
+		</buildCommand>
+		<buildCommand>
+			<name>org.eclipse.pde.ManifestBuilder</name>
```
Comment (AxelRICHARD) [inline packages/forms/backend/sirius-components-widget-reference-view-edit/.project]: I don't think this builder is needed.

File: packages/forms/backend/sirius-components-widget-reference-view-edit/.project [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+			<arguments>
+			</arguments>
+		</buildCommand>
+		<buildCommand>
+			<name>org.eclipse.pde.SchemaBuilder</name>
```
Comment (AxelRICHARD) [inline packages/forms/backend/sirius-components-widget-reference-view-edit/.project]: I don't think this builder is needed.

File: packages/forms/backend/sirius-components-widget-reference-view-edit/.project [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+			<arguments>
+			</arguments>
+		</buildCommand>
+	</buildSpec>
+	<natures>
+		<nature>org.eclipse.jdt.core.javanature</nature>
+		<nature>org.eclipse.m2e.core.maven2Nature</nature>
+		<nature>org.eclipse.pde.PluginNature</nature>
```
Comment (AxelRICHARD) [inline packages/forms/backend/sirius-components-widget-reference-view-edit/.project]: I don't think this nature is needed.

File: packages/forms/backend/sirius-components-widget-reference-view/.project [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+		<buildCommand>
+			<name>org.eclipse.pde.SchemaBuilder</name>
+			<arguments>
+			</arguments>
+		</buildCommand>
+	</buildSpec>
+	<natures>
+		<nature>org.eclipse.jdt.core.javanature</nature>
+		<nature>org.eclipse.m2e.core.maven2Nature</nature>
+		<nature>org.eclipse.pde.PluginNature</nature>
```
Comment (AxelRICHARD) [inline packages/forms/backend/sirius-components-widget-reference-view/.project]: I don't think this nature is needed.

File: packages/forms/backend/sirius-components-widget-reference-view/.project [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+			<arguments>
+			</arguments>
+		</buildCommand>
+		<buildCommand>
+			<name>org.eclipse.m2e.core.maven2Builder</name>
+			<arguments>
+			</arguments>
+		</buildCommand>
+		<buildCommand>
+			<name>org.eclipse.pde.ManifestBuilder</name>
```
Comment (AxelRICHARD) [inline packages/forms/backend/sirius-components-widget-reference-view/.project]: I don't think this builder is needed.

File: packages/forms/backend/sirius-components-widget-reference-view/.project [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+			<arguments>
+			</arguments>
+		</buildCommand>
+		<buildCommand>
+			<name>org.eclipse.pde.SchemaBuilder</name>
```
Comment (AxelRICHARD) [inline packages/forms/backend/sirius-components-widget-reference-view/.project]: I don't think this builder is needed.

File: packages/forms/frontend/sirius-components-widget-reference/src/SingleValuedReferencePropertySection.tsx [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+              </IconButton>
+            </ListItem>
+          ) : (
+            <ListItem>
+              <ListItemText data-testid="reference-value-none">None</ListItemText>
+            </ListItem>
+          )}
+        </List>
+      </div>
+      <Snackbar
```
Comment (AxelRICHARD) [inline packages/forms/frontend/sirius-components-widget-reference/src/SingleValuedReferencePropertySection.tsx]: The other widgets  (select, textfield...) don't use snackbar anymore, why this one (and  other reference property sections) still needs it?

File: packages/forms/frontend/sirius-components-widget-reference/src/ReferencePropertySection.tsx [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+  payload.__typename === 'SuccessPayload';
+
+export const ReferencePropertySection = ({
+  editingContextId,
+  formId,
+  widget,
+  subscribers,
+  readOnly,
+}: PropertySectionComponentProps<GQLReferenceWidget>) => {
+  const [message, setMessage] = useState<string | null>(null);
```
Comment (frouene) [inline packages/forms/frontend/sirius-components-widget-reference/src/ReferencePropertySection.tsx]: This state is never used and unnecessary

File: packages/forms/frontend/sirius-components-widget-reference/src/ReferencePropertySection.tsx [hunk 10]
```diff
@@ -0,0 +1,269 @@
...
+        onClose={() => setMessage(null)}
+        message={message}
+        action={
+          <IconButton size="small" aria-label="close" color="inherit" onClick={() => setMessage(null)}>
+            <CloseIcon fontSize="small" />
+          </IconButton>
+        }
+        data-testid="error"
+      />
+    </div>
```
Comment (frouene) [inline packages/forms/frontend/sirius-components-widget-reference/src/ReferencePropertySection.tsx]: Snackbar must not be used in widget component anymore

File: packages/forms/frontend/sirius-components-widget-reference/src/SingleValuedReferencePropertySection.tsx [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+  payload.__typename === 'SuccessPayload';
+
+export const SingleValuedReferencePropertySection = ({
+  editingContextId,
+  formId,
+  widget,
+  subscribers,
+  readOnly,
+}: PropertySectionComponentProps<GQLSingleValuedReferenceWidget>) => {
+  const [message, setMessage] = useState<string | null>(null);
```
Comment (frouene) [inline packages/forms/frontend/sirius-components-widget-reference/src/SingleValuedReferencePropertySection.tsx]: This state is never used and unnecessary

File: packages/forms/frontend/sirius-components-widget-reference/src/SingleValuedReferencePropertySection.tsx [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+        onClose={() => setMessage(null)}
+        message={message}
+        action={
+          <IconButton size="small" aria-label="close" color="inherit" onClick={() => setMessage(null)}>
+            <CloseIcon fontSize="small" />
+          </IconButton>
+        }
+        data-testid="error"
+      />
+    </div>
```
Comment (frouene) [inline packages/forms/frontend/sirius-components-widget-reference/src/SingleValuedReferencePropertySection.tsx]: Snackbar must not be used in widget component anymore

File: packages/forms/frontend/sirius-components-widget-reference/vite.config.js [hunk 13]
```diff
@@ -0,0 +1,23 @@
+import react from '@vitejs/plugin-react';
+import path from 'node:path';
+import peerDepsExternal from 'rollup-plugin-peer-deps-external';
+import { defineConfig } from 'vite';
+
+export default defineConfig({
+  plugins: [peerDepsExternal(), react()],
+  build: {
+    lib: {
+      name: 'sirius-components-forms',
```
Comment (frouene) [inline packages/forms/frontend/sirius-components-widget-reference/vite.config.js]: The name should be 'sirius-components-widget-reference' ?

File: packages/forms/backend/sirius-components-widget-reference/src/main/java/org/eclipse/sirius/components/widget/reference/ReferenceElementProps.java [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+    }
+
+    @Override
+    public String toString() {
+        String pattern = "{0} '{'id: {1}, label: {2}'}'";
+        return MessageFormat.format(pattern, this.getClass().getSimpleName(), this.id, this.label);
+    }
+
+    /**
+     * The builder of the slider element props.
```
Comment (frouene) [inline packages/forms/backend/sirius-components-widget-reference/src/main/java/org/eclipse/sirius/components/widget/reference/ReferenceElementProps.java]: it doesn't concern the slider

File: packages/forms/backend/sirius-components-widget-reference/src/main/java/org/eclipse/sirius/components/widget/reference/ReferenceWidgetDescriptionConverterProvider.java [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+import org.eclipse.sirius.components.core.api.IEditService;
+import org.eclipse.sirius.components.core.api.IFeedbackMessageService;
+import org.eclipse.sirius.components.core.api.IObjectService;
+import org.eclipse.sirius.components.forms.description.AbstractWidgetDescription;
+import org.eclipse.sirius.components.interpreter.AQLInterpreter;
+import org.eclipse.sirius.components.view.emf.form.IWidgetConverterProvider;
+import org.springframework.stereotype.Service;
+
+/**
+ * Provides the widget converter needed for the Slider widget.
```
Comment (frouene) [inline packages/forms/backend/sirius-components-widget-reference/src/main/java/org/eclipse/sirius/components/widget/reference/ReferenceWidgetDescriptionConverterProvider.java]: it doesn't concern the slider

File: packages/forms/backend/sirius-components-widget-reference-view-edit/src/main/resources/plugin.properties [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+_UI_CreateChild_tooltip = Create New {0} Under {1} Feature
+_UI_CreateChild_description = Create a new child of type {0} for the {1} feature of the selected {2}.
+_UI_CreateSibling_description = Create a new sibling of type {0} for the selected {2}, under the {1} feature of their parent.
+
+_UI_PropertyDescriptor_description = The {0} of the {1}
+
+_UI_ReferenceWidgetDescription_type = Reference Widget Description
+_UI_Unknown_type = Object
+
+_UI_Unknown_datatype= Value
```
Comment (frouene) [inline packages/forms/backend/sirius-components-widget-reference-view-edit/src/main/resources/plugin.properties]: a space is missing before =

File: package-lock.json
Comment (pcdavid) [reply package-lock.json:1]: I think yes, because there is a new frontend package (`packages/forms/frontend/sirius-components-widget-reference`).

**Review by frouene — CHANGES_REQUESTED** [review]

File: packages/formdescriptioneditors/frontend/sirius-components-formdescriptioneditors/src/WidgetEntry.tsx (hunk 1 above)
Comment (pcdavid) [reply packages/formdescriptioneditors/frontend/sirius-components-formdescriptioneditors/src/WidgetEntry.tsx:506]: I don't really understand why, but it's the only way I found to allow the "Delete" key (more precisely, the `onKeyDown` handler) to work with the new widget. Based on https://bobbyhadz.com/blog/react-onkeydown-div.

File: doc/adrs/105_reference_widget.adoc [hunk 17]
```diff
@@ -0,0 +1,131 @@
...
+
+* To clear/unset a single or multi-valued reference, use `newValueIds = []`
+* To set the value of a single-valued reference, use `newValueIds = [ semanticIdOfTheNewValue ]`
+* To add, remove or re-order elements in a multi-valued reference, simply perform the addition/removal/reorder on the frontend using the ids of the current `ReferenceValue`, and invoke the mutation with the resulting ids, in the expected order, as `newValueIds`.
+
+The `EditReferenceEventHandler` (which implements `IFormEventHandler`) on the backend will be responsible to interpret and validate the `newValueIds` it receives in light of the characteristics of the actual _EReference_, and return any meaningful error message.
+This includes making sur the actual objects referenced by the `newValueIds` are type-compatible with the reference.
+
+On the frontend, the `ReferenceWidget` will be provided in a separate package.
+As was the was for the Slider custom widget, the Sirius Web Sample application will depend on it and declare its metadata in the top-level `propertySectionsRegistry`.
```
Comment (frouene) [inline doc/adrs/105_reference_widget.adoc]: Typo: As was the was

File: packages/forms/backend/sirius-components-widget-reference-view-edit/.project (hunk 2 above)
Comment (pcdavid) [reply packages/forms/backend/sirius-components-widget-reference-view-edit/.project]: done

File: packages/forms/backend/sirius-components-widget-reference-view-edit/.project (hunk 3 above)
Comment (pcdavid) [reply packages/forms/backend/sirius-components-widget-reference-view-edit/.project]: done

File: packages/forms/backend/sirius-components-widget-reference-view-edit/.project (hunk 4 above)
Comment (pcdavid) [reply packages/forms/backend/sirius-components-widget-reference-view-edit/.project]: done

File: packages/forms/backend/sirius-components-widget-reference-view/.project (hunk 6 above)
Comment (pcdavid) [reply packages/forms/backend/sirius-components-widget-reference-view/.project]: done

File: packages/forms/backend/sirius-components-widget-reference-view/.project (hunk 5 above)
Comment (pcdavid) [reply packages/forms/backend/sirius-components-widget-reference-view/.project]: done

File: packages/forms/backend/sirius-components-widget-reference-view/.project (hunk 7 above)
Comment (pcdavid) [reply packages/forms/backend/sirius-components-widget-reference-view/.project]: done

File: packages/forms/frontend/sirius-components-widget-reference/src/SingleValuedReferencePropertySection.tsx (hunk 8 above)
Comment (pcdavid) [reply packages/forms/frontend/sirius-components-widget-reference/src/SingleValuedReferencePropertySection.tsx]: This file is actually dead code that I forgot to delete.
The single widget/property section if in `ReferencePropertySection`, which indeed has the same issue, and I will fix it there.

File: packages/forms/frontend/sirius-components-widget-reference/src/ReferencePropertySection.tsx (hunk 10 above)
Comment (pcdavid) [reply packages/forms/frontend/sirius-components-widget-reference/src/ReferencePropertySection.tsx]: Done

File: packages/forms/frontend/sirius-components-widget-reference/src/ReferencePropertySection.tsx (hunk 9 above)
Comment (pcdavid) [reply packages/forms/frontend/sirius-components-widget-reference/src/ReferencePropertySection.tsx]: Done.

File: packages/forms/frontend/sirius-components-widget-reference/src/SingleValuedReferencePropertySection.tsx (hunk 11 above)
Comment (pcdavid) [reply packages/forms/frontend/sirius-components-widget-reference/src/SingleValuedReferencePropertySection.tsx:70]: Done.

Comment (pcdavid) [reply packages/forms/frontend/sirius-components-widget-reference/src/SingleValuedReferencePropertySection.tsx]: Actually the whole file is dead code that I forgot to remove.

File: packages/forms/frontend/sirius-components-widget-reference/vite.config.js (hunk 13 above)
Comment (pcdavid) [reply packages/forms/frontend/sirius-components-widget-reference/vite.config.js]: Yes. Done.

File: packages/forms/backend/sirius-components-widget-reference/src/main/java/org/eclipse/sirius/components/widget/reference/ReferenceWidgetDescriptionConverterProvider.java (hunk 15 above)
Comment (pcdavid) [reply packages/forms/backend/sirius-components-widget-reference/src/main/java/org/eclipse/sirius/components/widget/reference/ReferenceWidgetDescriptionConverterProvider.java]: fixed.

File: packages/forms/backend/sirius-components-widget-reference/src/main/java/org/eclipse/sirius/components/widget/reference/ReferenceElementProps.java (hunk 14 above)
Comment (pcdavid) [reply packages/forms/backend/sirius-components-widget-reference/src/main/java/org/eclipse/sirius/components/widget/reference/ReferenceElementProps.java]: fixed.

File: packages/forms/backend/sirius-components-widget-reference-view-edit/src/main/resources/plugin.properties (hunk 16 above)
Comment (pcdavid) [reply packages/forms/backend/sirius-components-widget-reference-view-edit/src/main/resources/plugin.properties]: done

File: doc/adrs/105_reference_widget.adoc (hunk 17 above)
Comment (pcdavid) [reply doc/adrs/105_reference_widget.adoc]: fixed.

**Review by AxelRICHARD — APPROVED** [review]

**Pull request merged**